File: py_scripts/Sensor.py
import numpy as np
from DynamicBayesianFiltering import SensorData

class Sensor:
    sensor_radius: float # Length of each ray
    sensor_angle: float # Angle about the x-axis dictating the field of view
    num_rays: int # Number of rays to send out

    def __init__(self, sensor_radius: float, sensor_angle: float, num_rays: int):
        self.sensor_radius = sensor_radius
        self.sensor_angle = sensor_angle
        self.num_rays = num_rays

    # Check if a ray sent from the current position collided with an obstacle
    # returns true if a collision happened, false otherwise
    def __check_collision(self, obstacles: np.ndarray, end_point: np.ndarray, pos_c: np.ndarray, collision_threshold = 1):
        # Number of steps to check along for collision
        # [5.1302186 4.0434062 0.
        num_steps = int(np.linalg.norm(end_point - pos_c[:-1]))
        for step in range(1, num_steps + 1):
            
            # Calc the point to check
            point_c = pos_c[:-1] + (end_point - pos_c[:-1]) * (step / num_steps)

            for obs in obstacles:
                if pos_c[0] - 5.1302186 <= 1e-2 and np.linalg.norm(point_c - obs[:-1]) < 1.7 and False:
                    print(np.linalg.norm(pos_c[:-1] - obs[:-1]))
                    #print("endpoint", np.linalg.norm(end_point - pos_c[:-1]))
                if np.linalg.norm(point_c - obs[:-1]) < collision_threshold:
                    return True
            
        return False

    # Scan the area around the current position to detect either occupied points or free space
    # returns a list of rays with the end point and whether a collision happened
    def scan_area(self, pos_c: np.ndarray, obstacles: np.ndarray):
        angle_increment = self.sensor_angle / self.num_rays
        rays = []

        cos_vals = np.cos(np.radians(np.linspace(-self.sensor_angle / 2, self.sensor_angle / 2, self.num_rays)))
        sin_vals = np.sin(np.radians(np.linspace(-self.sensor_angle / 2, self.sensor_angle / 2, self.num_rays)))

        for x, y in zip(cos_vals, sin_vals):
            direction = np.array([x, y])
            direction = direction / np.linalg.norm(direction)
            end_point = direction * self.sensor_radius + pos_c[:-1]

            collision = self.__check_collision(obstacles, end_point, pos_c)
            rays.append([end_point, collision])

        return rays
    
    # Returns the area scanned by the sensor
    def calc_scanned_area(self):
        return np.pi * self.sensor_radius**2 * self.sensor_angle / 360
    
    # Returns the AOI (Area of Interest)
    def calc_area_of_interest(self, pos_c: np.ndarray, local_min: np.ndarray):
        # Calculate the distance between the UGV and the local minimum
        distance = np.linalg.norm(local_min - pos_c)

        current_area = self.calc_scanned_area()
        area_from_local_min = np.pi * distance**2 * self.sensor_angle / 360

        return current_area + area_from_local_min
    
    # Returns  the obstacles in the recognized/scanned area
    def get_obstacles_in_recognized_area(self, pos_c: np.ndarray, obstacles: np.ndarray):
        
        seen_obstacles = []
        for obs in obstacles:
            if np.linalg.norm(obs - pos_c) < self.sensor_radius:
                seen_obstacles.append(obs)
        
        return np.array(seen_obstacles)
            
    # Returns the number of points in group A
    # "Group A comprises points between the initial local minimum and the UGV, including the local minimum itself" (from research paper)
    def get_group_a_points(self, pos_c: np.ndarray, local_min: np.ndarray, step_size = 0.1):
        distance = np.linalg.norm(local_min - pos_c)
        return int(np.ceil(distance / step_size)) # np.ceil automatically rounds up the number no matter the decimal value (e.g. 0.2 -> 1)
    
    # Returns true if the forces are parrallel, false otherwise
    def check_for_parrallel_forces(self, pos_c: np.ndarray, F_att: np.ndarray, F_rep: np.ndarray):
       
        # Check if the forces are parrallel
        cross_product = np.cross(F_att, F_rep)
        cross_product = cross_product[:-1]
        return np.allclose(cross_product, 0)
    
    # Returns the data from the sensor for the Dynamic Bayesian Filtering Algorithm
    # Returns in the form of SensorData from DynamicBayesianFiltering.py
    def get_sensor_data(self, pos_c: np.ndarray, local_min: np.ndarray, obstacles: np.ndarray):
        rays = self.scan_area(pos_c, obstacles)
        
        # Count the number of rays that detected a collision
        num_collisions = sum(1 for ray in rays if ray[1])
        recognized_area = self.calc_scanned_area()
        total_area = self.calc_area_of_interest(pos_c, local_min)
        
        return SensorData(num_collisions, len(rays), recognized_area, total_area)
        

File: cpp_code/Sensor.cpp
#include "Sensor.hpp"
#include <cmath>
#define _USE_MATH_DEFINES
#include <math.h>
#include <algorithm>
#include <sstream>
#include <iostream>
#include "DynamicBayesianFiltering.hpp"

// Constructor
Sensor::Sensor(double sensor_radius, double sensor_angle, int num_rays)
    : sensor_radius(sensor_radius), sensor_angle(sensor_angle), num_rays(num_rays) {}

// Check for collisions along a path
bool Sensor::check_collision(const std::vector<std::array<double, 3>>& obstacles, const std::array<double, 2>& end_point, const std::array<double, 3>& pos_c, double collision_threshold) {
    int num_steps = static_cast<int>(std::sqrt(std::pow(end_point[0] - pos_c[0], 2) + std::pow(end_point[1] - pos_c[1], 2)));
    for (int step = 1; step <= num_steps; ++step) {
        std::array<double, 2> point_c = { pos_c[0] + (end_point[0] - pos_c[0]) * (step / static_cast<double>(num_steps)),
                                          pos_c[1] + (end_point[1] - pos_c[1]) * (step / static_cast<double>(num_steps)) };
        for (const auto& obs : obstacles) {
            if (std::sqrt(std::pow(point_c[0] - obs[0], 2) + std::pow(point_c[1] - obs[1], 2)) < collision_threshold) {
                return true;
            }
        }
    }
    return false;
}

// Scan the area and detect obstacles
std::vector<std::pair<std::array<double, 2>, bool>> Sensor::scan_area(const std::array<double, 3>& pos_c, const std::vector<std::array<double, 3>>& obstacles) {
    double angle_increment = sensor_angle / num_rays;
    std::vector<std::pair<std::array<double, 2>, bool>> rays;

    for (int i = 0; i < num_rays; ++i) {
        double angle = -sensor_angle / 2 + i * angle_increment;
        std::array<double, 2> direction = { std::cos(angle), std::sin(angle) };
        std::array<double, 2> end_point = { pos_c[0] + direction[0] * sensor_radius, pos_c[1] + direction[1] * sensor_radius };
        bool collision = check_collision(obstacles, end_point, pos_c);
        rays.push_back({ end_point, collision });
    }

    return rays;
}

// Calculate the scanned area
double Sensor::calc_scanned_area() {
    return M_PI * std::pow(sensor_radius, 2) * (sensor_angle / 360.0);
}

// Calculate the area of interest
double Sensor::calc_area_of_interest(const std::array<double, 3>& pos_c, const std::array<double, 3>& local_min) {
    double distance = std::sqrt(std::pow(local_min[0] - pos_c[0], 2) + std::pow(local_min[1] - pos_c[1], 2));
    double current_area = calc_scanned_area();
    double area_from_local_min = M_PI * std::pow(distance, 2) * (sensor_angle / 360.0);
    return current_area + area_from_local_min;
}

// Get obstacles within the sensor's recognized area
std::vector<std::array<double, 3>> Sensor::get_obstacles_in_recognized_area(const std::array<double, 3>& pos_c, const std::vector<std::array<double, 3>>& obstacles) {
    std::vector<std::array<double, 3>> seen_obstacles;
    for (const auto& obs : obstacles) {
        double distance = std::sqrt(std::pow(obs[0] - pos_c[0], 2) + std::pow(obs[1] - pos_c[1], 2));
        if (distance < sensor_radius) {
            seen_obstacles.push_back(obs);
        }
    }
    return seen_obstacles;
}

// Get the number of points in group A
int Sensor::get_group_a_points(const std::array<double, 3>& pos_c, const std::array<double, 3>& local_min, double step_size) {
    double distance = std::sqrt(std::pow(local_min[0] - pos_c[0], 2) + std::pow(local_min[1] - pos_c[1], 2));
    return static_cast<int>(std::ceil(distance / step_size));
}

// Check if forces are parallel
bool Sensor::check_for_parallel_forces(const std::array<double, 3>& F_att, const std::array<double, 3>& F_rep) {
    std::array<double, 3> cross_product = { F_att[1] * F_rep[2] - F_att[2] * F_rep[1],
                                            F_att[2] * F_rep[0] - F_att[0] * F_rep[2],
                                            F_att[0] * F_rep[1] - F_att[1] * F_rep[0] };
    return std::all_of(cross_product.begin(), cross_product.end(), [](double val) { return std::abs(val) < 1e-5; });
}

// Get sensor data and log it
SensorData Sensor::get_sensor_data(const std::array<double, 3>& pos_c, const std::array<double, 3>& local_min, const std::vector<std::array<double, 3>>& obstacles, std::string& log_str) {
    auto rays = scan_area(pos_c, obstacles);
    int num_collisions = std::count_if(rays.begin(), rays.end(), [](const auto& ray) { return ray.second; });
    double recognized_area = calc_scanned_area();
    double total_area = calc_area_of_interest(pos_c, local_min);

    std::ostringstream oss;
    oss << "Num Collisions: " << num_collisions << ", Recognized Area: " << recognized_area << ", Total Area: " << total_area;
    log_str = oss.str();

    return { num_collisions, static_cast<int>(rays.size()), recognized_area, total_area };
}